### PR TITLE
[pmm-agent] PMM-6256 Fixed MongoDB explain.

### PR DESCRIPTION
**Pull request by percona-csalguero**
Previously, the db used for explain was always "admin" but for some
query types we need to use the db we have in the query example.

### CODE REVIEW

File: actions/aggregate.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{"ns":"sbtest.orders","op":"command","command":{"aggregate":"orders","pipeline":[{"$match":{"status":"A"}},{"$group":{"_id":"$cust_id","total":{"$sum":"$amount"}}},{"$sort":{"total":-1}}],"cursor":{},"$db":"sbtest"}}
```
Comment (AlekSi) [inline actions/aggregate.json:1]: I guess we should not have those files outside of testdata?

Comment (BupycHuk) [reply actions/aggregate.json:1]: +1

File: actions/mongodb_explain_action_test.go [hunk 2]
```diff
@@ -19,11 +19,17 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
 	"testing"
 
+	"github.com/Masterminds/semver"
```
Comment (AlekSi) [inline actions/mongodb_explain_action_test.go:27]: We have our own package for that: https://github.com/percona/pmm/blob/PMM-2.0/version/parsed.go

Example of usage:
* https://github.com/percona/pmm-managed/blob/22524d05feea287a2d315df3e5c2f557242ef9e4/services/checks/checks.go#L74-L79
* https://github.com/percona/pmm-managed/blob/22524d05feea287a2d315df3e5c2f557242ef9e4/services/checks/checks.go#L640

File: actions/mongodb_explain_action_test.go [hunk 3]
```diff
@@ -74,6 +79,82 @@ func TestMongoDBExplain(t *testing.T) {
...
+	testFiles := []struct {
+		in         string
+		constraint string
+	}{
+		{
+			in: "distinct.json",
+		},
+		{
+			in:         "aggregate.json",
+			constraint: "3.6.0",
```
Comment (AlekSi) [inline actions/mongodb_explain_action_test.go:106]: Is that minimal version? Then I think we should call it `minVersion` or something. That should also simplify generic SemVer comparison function below.

Comment (BupycHuk) [inline actions/mongodb_explain_action_test.go:106]: We have mongo:3.4 in the matrix,so I think it should be changed to pass tests.

Comment (BupycHuk) [reply actions/mongodb_explain_action_test.go:106]: +1

File: actions/mongodb_explain_action_test.go [hunk 4]
```diff
@@ -89,3 +170,25 @@ func prepareData(ctx context.Context, client *mongo.Client, database, collection
 
 	return nil
 }
+
+func notLessThan(minVersionStr, vin string) (bool, error) {
```
Comment (BupycHuk) [inline actions/mongodb_explain_action_test.go:174]: vin?

File: agents/mongodb/internal/profiler/profiler_test.go [hunk 5]
```diff
@@ -207,20 +207,20 @@ func TestProfiler(t *testing.T) {
 		assert.Nil(t, err)
 
 		want := map[string]interface{}{
-			"indexFilterSet": bool(false),
+			"indexFilterSet": false,
+			"namespace":      "test_00.people",
 			"parsedQuery": map[string]interface{}{
 				"name_00�": map[string]interface{}{
```
Comment (BupycHuk) [inline agents/mongodb/internal/profiler/profiler_test.go:213]: let's use `\xfdd` type vaues instead of `�`

Comment (AlekSi) [reply agents/mongodb/internal/profiler/profiler_test.go:213]: ^

File: actions/testdata/find_and_modify.json [hunk 6]
```diff
@@ -0,0 +1 @@
+{"ns":"sbtest.people","op":"command","command":{"findandmodify":"people","query":{"name":"Andy"},"sort":{"rating":1},"update":{"$inc":{"score":1}},"upsert":true,"$db":"sbtest"}}
```
Comment (AlekSi) [inline actions/testdata/find_and_modify.json:1]: Let's add \n

File: actions/mongodb_explain_action_test.go [hunk 7]
```diff
@@ -89,3 +170,25 @@ func prepareData(ctx context.Context, client *mongo.Client, database, collection
...
+	// Drop everything after first dash.
+	// Version with dash is considered a pre-release
+	// but some MongoDB builds add additional information after dash
+	// even though it's not considered a pre-release but a release.
```
Comment (AlekSi) [inline actions/mongodb_explain_action_test.go:178]: I think we don't need that. Just call `version.Parse`, then `v.Rest = ""`

File: actions/mongodb_explain_action_test.go (hunk 4 above)
Comment (percona-csalguero) [reply actions/mongodb_explain_action_test.go:174]: inVersion. Changed for clarity

**Comment by AlekSi** [comment]
@BupycHuk Please move ticket if you decline PR

**Comment by percona-csalguero** [comment]
@BupycHuk All tests are passing.

File: agents/mongodb/internal/profiler/profiler_test.go [hunk 8]
```diff
@@ -105,15 +105,15 @@ func TestProfiler(t *testing.T) {
 		fieldsCount := dbNumber + 1
 		doc := bson.M{}
 		for j := 0; j < fieldsCount; j++ {
-			doc[fmt.Sprintf("name_%02d\xff", j)] = fmt.Sprintf("value_%02d\xff", j) // to generate different fingerprints
```
Comment (BupycHuk) [inline agents/mongodb/internal/profiler/profiler_test.go:108]: this was done for the reason. Why did you remove it?

Comment (percona-csalguero) [reply agents/mongodb/internal/profiler/profiler_test.go:108]: Didn't know it was an UTF8 test. I thought it was a copy of a random data generator.
Restored the test.

File: agents/mongodb/internal/profiler/profiler_test.go (hunk 5 above)
Comment (BupycHuk) [reply agents/mongodb/internal/profiler/profiler_test.go:213]: @percona-csalguero could you try to replace it with `\ufffd`

File: Gopkg.toml [hunk 9]
```diff
@@ -39,7 +39,7 @@ required = [
 [[constraint]]
   name = "github.com/percona/percona-toolkit"
   # version = "3.1.1"
-  branch = "PMM-4192_Mongo_explain_fails_fixing_tests"
+  branch = "PMM-6256_Explain_doesnt_work_2"
```
Comment (askomorokhov) [inline Gopkg.toml:42]: 👀

File: actions/testdata/aggregate.json [hunk 10]
```diff
@@ -0,0 +1 @@
+{"ns":"sbtest.orders","op":"command","command":{"aggregate":"orders","pipeline":[{"$match":{"status":"A"}},{"$group":{"_id":"$cust_id","total":{"$sum":"$amount"}}},{"$sort":{"total":-1}}],"cursor":{},"$db":"sbtest"}}
```
Comment (askomorokhov) [inline actions/testdata/aggregate.json:1]: Can we please format this file?

File: vendor/github.com/percona/percona-toolkit/src/go/mongolib/proto/system.profile.go [hunk 11]
```diff
@@ -178,7 +178,7 @@ func (self ExampleQuery) ExplainCmd() bson.D {
 					} else {
 						cmd = append(cmd[:i], cmd[i+1:]...)
 					}
-					break
+					// break LOOP
```
Comment (askomorokhov) [inline vendor/github.com/percona/percona-toolkit/src/go/mongolib/proto/system.profile.go:181]: Was this `break` removed with intention?

Comment (percona-csalguero) [reply vendor/github.com/percona/percona-toolkit/src/go/mongolib/proto/system.profile.go:181]: yes. it was unused and highlighted by the linter

File: actions/testdata/aggregate.json (hunk 10 above)
Comment (percona-csalguero) [reply actions/testdata/aggregate.json]: Done

File: Gopkg.toml (hunk 9 above)
Comment (percona-csalguero) [reply Gopkg.toml:42]: There is new code in toolkit for this.

File: actions/testdata/aggregate.json (hunk 10 above)
Comment (askomorokhov) [reply actions/testdata/aggregate.json]: 🍻

File: Gopkg.toml (hunk 9 above)
Comment (AlekSi) [reply Gopkg.toml:42]: What @askomorokhov means is that we should not merge the change with that branch

Comment (AlekSi) [reply Gopkg.toml:42]: https://github.com/percona/percona-toolkit/pull/459 was merged, so we should update a branch
